File: src/main/java/downloader/LauncherDownloader.kt
package downloader

import DirectoryHelper
import com.google.gson.Gson
import sk.tomsik68.mclauncher.api.ui.IProgressMonitor
import sk.tomsik68.mclauncher.util.FileUtils
import sk.tomsik68.mclauncher.util.HttpUtils
import utils.Utils
import java.io.File

private const val LAUNCHER_JSON_URL = "https://minecraft.glitchless.ru/launcher.json"

class LauncherDownloader {
    val gson = Gson()
    var launcherModel: LauncherModel? = null

    fun initDownloader() {
        try {
            val json = HttpUtils.httpGet(LAUNCHER_JSON_URL)
            launcherModel = gson.fromJson(json, LauncherModel::class.java)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun checkFile(): Boolean {
        if(!DirectoryHelper.getLauncherFile().exists()) {
            return false
        }

        if (launcherModel == null) {
            return true
        }

        val sha256 = Utils.generateSHA256(DirectoryHelper.getLauncherFile())
        return sha256 == launcherModel!!.sha256
    }

    fun updateLauncher(monitor: IProgressMonitor)  {
        if (launcherModel == null) {
            return
        }

        monitor.setStatus("Downloading launcher...")
        val updateFile = File(DirectoryHelper.getTemporaryDirectory(), "update_launcher.jar");
        FileUtils.downloadFileWithProgress(launcherModel!!.downloadUrl, updateFile, monitor)

        if (Utils.generateSHA256(updateFile) != launcherModel!!.sha256) {
            return
        }

        if (!DirectoryHelper.getLauncherFile().exists() || DirectoryHelper.getLauncherFile().delete()) {
            updateFile.renameTo(DirectoryHelper.getLauncherFile())
        }
    }
}

File: src/net/LauncherDownloader.cpp
#include "net/LauncherDownloader.h"
#include "net/Downloader.h"
#include "app/ProgressMonitor.h"
#include "util/Paths.h"
#include "util/Hash.h"
#include <QFile>
#include <QFileInfo>
#include <QObject>

namespace tprunner {

namespace { const char* kLauncherUrl = "https://minecraft.glitchless.ru/launcher.json"; }

LauncherDownloader::LauncherDownloader(Downloader* dl) : dl_(dl) {}

QByteArray LauncherDownloader::fetchManifest() { return dl_->httpGet(kLauncherUrl); }

void LauncherDownloader::init() {
    try { model_ = LauncherModel::fromJson(fetchManifest()); }
    catch (...) { model_ = std::nullopt; }
}

QString LauncherDownloader::jreCode() const { return model_ ? model_->jreCode : QString(); }
QList<JavaBinaryModel> LauncherDownloader::jreFiles() const { return model_ ? model_->jreFiles : QList<JavaBinaryModel>(); }

QString LauncherDownloader::selectedJarUrl() const {
    if (!model_) return QString();
    if (const auto m = selectForCurrentPlatform(model_->files)) return m->downloadUrl;
    return model_->downloadUrl;
}

QString LauncherDownloader::selectedJarSha() const {
    if (!model_) return QString();
    if (const auto m = selectForCurrentPlatform(model_->files)) return m->sha256;
    return model_->sha256;
}

bool LauncherDownloader::checkFile() {
    if (!QFileInfo::exists(Paths::launcherFile())) return false;
    if (!model_) return true;                       // can't verify → assume OK (matches original)
    return sha256Base64(Paths::launcherFile()) == selectedJarSha();
}

void LauncherDownloader::update(ProgressMonitor* monitor) {
    if (!model_) return;
    const QString url = selectedJarUrl();
    if (url.isEmpty()) return;
    if (monitor) monitor->setStatus(QObject::tr("Скачивание лаунчера..."));
    const QString updateFile = Paths::temporaryDirectory() + "/update_launcher.jar";
    dl_->downloadToFile(url, updateFile, monitor);
    if (monitor) monitor->setProgress(100);

    const QString sha = selectedJarSha();
    if (!sha.isEmpty() && sha256Base64(updateFile) != sha) return;   // verification failed → keep old jar

    QFile launcher(Paths::launcherFile());
    if (!launcher.exists() || launcher.remove())
        QFile::rename(updateFile, Paths::launcherFile());
}

}
